File: SORT/sort_ian.py

class BED:
	def __init__(self, chr, beg, end):
		self.chr = chr
		self.beg = beg
		self.end = end
	def __repr__(self):
		return repr((self.chr, self.beg, self.end))
	def length(self):
		return 'CBA'.index(self.end) - (self.beg)


bedfile = [
	BED('A', 100, 300),
	BED('B', 100, 200),
	BED('A', 100, 200),
	BED('A', 200, 300),
]

print("by chrom")
beds = sorted(bedfile, key=lambda bed: (bed.chr, bed.beg, bed.end))
print(beds)

File: SORT/sort_ian.pl
#!/usr/bin/perl
use strict;
use warnings FATAL => 'all';

my @bed;
while (<>) {
	chomp;
	my ($chr, $beg, $end) = split;
	push @bed, {
		'chr' => $chr,
		'beg' => $beg,
		'end' => $end,
	};
}

print "by chrom\n";
@bed = sort {$a->{'chr'} cmp $b->{'chr'}} @bed;
printbed(\@bed);

print "by length\n";
@bed = sort {($a->{'end'} - $a->{'beg'}) <=> ($b->{'end'} - $b->{'beg'})} @bed;
printbed(\@bed);

print "by chr,beg,end\n";
@bed = sort {$a->{'chr'} cmp $b->{'chr'} or $a->{beg} <=> $b->{beg} or $a->{end} <=> $b->{end}} @bed;
printbed(\@bed);

sub printbed {
	my ($bed) = @_;
	foreach my $f (@$bed) {
		print "$f->{chr}\t$f->{beg}\t$f->{end}\n";
	}
}
